tests/src/test/java/org/sonarqube/tests/rule/RuleReKeyingTest.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2018 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarqube.tests.rule;
+
+import com.sonar.orchestrator.Orchestrator;
+import java.io.File;
+import org.junit.After;
+import org.junit.Test;
+import org.sonarqube.qa.util.Tester;
+import org.sonarqube.ws.client.PostRequest;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static util.ItUtils.pluginArtifact;
+
+public class RuleReKeyingTest {
+
+  private static Orchestrator orchestrator;
+  private static Tester tester;
+
+  @After
+  public void tearDown() {
+    if (tester != null) {
+      tester.after();
+    }
+    if (orchestrator != null) {
+      orchestrator.stop();
+    }
+  }
+
+  @Test
+  public void rules_are_re_keyed_when_upgrading_and_downgrading_plugin() {
+    orchestrator = Orchestrator.builderEnv()
+      .addPlugin(pluginArtifact("foo-plugin-v1"))
+      .build();
+    orchestrator.start();
+
+    tester = new Tester(orchestrator);
+    tester.before();
+
+    verifyRuleCount(16, 16);
```

Comment:
Not easy to understand where those numbers comes from.
Maybe testing with a list of expected rules will make it clearer.